File: Chapter 04/JS/section_4_4_anti_patterns/after_rich_domain_thin_controller/domain/interfaces/dataAccessInterfaces.js
/**
 * ARCHITECTURE NOTE: In a traditional Layered Architecture, the 
 * Infrastructure Layer defines the contracts for accessing data. 
 * The Business Logic layer above will be forced to depend on 
 * this layer to use these interfaces.
 * 
 * * In JavaScript, we simulate interfaces by creating base classes 
 * that throw errors if their methods aren't overridden.
 */
class IOrderRepository {
    getById(orderId) { throw new Error("Not implemented"); }
    save(order) { throw new Error("Not implemented"); }
}

class ICustomerRepository {
    getById(customerId) { throw new Error("Not implemented"); }
}

class IEmailService {
    send(to, subject, body) { throw new Error("Not implemented"); }
}

class IItemRepository {
    /**
     * ARCHITECTURE NOTE: This is the "Security Hook" the Service uses 
     * to verify prices. Fetching the item from the database ensures 
     * we use the official price, not one sent in a request DTO. Defines the contract for data access.
     * The Application Layer depends on this interface, not a concrete DB.
     * * @param {number} itemId 
     * @returns {Item}
     */
    getById(itemId) {
        throw new Error("Method 'getById()' must be implemented.");
    }
}

module.exports = {
    IOrderRepository,
    ICustomerRepository,
    IEmailService,
    IItemRepository
};

File: Chapter 04/Python/section_4_4_anti_patterns/after_rich_domain_thin_controller/domain/interfaces/data_access_interfaces.py
from abc import ABC, abstractmethod
from ..models.item import Item

class IOrderRepository(ABC):
    """
    ARCHITECTURE NOTE: In a traditional N-Tier architecture, the 
    Data Access Layer defines the contracts for accessing data. 
    The Business Logic layer above will be forced to depend on 
    this layer to use these interfaces.
    """
    @abstractmethod
    def get_by_id(self, order_id: int):
        pass

    @abstractmethod
    def save(self, order):
        pass

class IItemRepository(ABC):
    @abstractmethod
    def get_by_id(self, item_id: int) -> Item:
        """This is the 'Security Hook' the Service uses to verify prices"""
        pass

class ICustomerRepository(ABC):
    @abstractmethod
    def get_by_id(self, customer_id: int):
        pass

class IEmailService(ABC):
    @abstractmethod
    def send(self, to: str, subject: str, body: str):
        pass